src/base/utils/string.h

```diff
@@ -36,19 +36,45 @@
 #include <QString>
 #include <Qt>
 #include <QtContainerFwd>
+#include <QtGlobal>
+
+#ifndef Q_OS_WIN
+#define QBT_USE_QCOLLATOR
+#include <QCollator>
+#endif
 
 class QStringRef;
 
 namespace Utils::String
 {
-    QString fromDouble(double n, int precision);
+#ifndef QBT_USE_QCOLLATOR
+    int naturalCompare(const QString &left, const QString &right, Qt::CaseSensitivity caseSensitivity);
+#endif
 
-    int naturalCompare(const QString &left, const QString &right, const Qt::CaseSensitivity caseSensitivity);
     template <Qt::CaseSensitivity caseSensitivity>
-    bool naturalLessThan(const QString &left, const QString &right)
+    class NaturalCompare
     {
-        return (naturalCompare(left, right, caseSensitivity) < 0);
-    }
+    public:
+#ifdef QBT_USE_QCOLLATOR
+        NaturalCompare();
```

Comment:
I would declare constructor unconditionally but rather use conditionals in its body.

Reply:
Really, I'm just annoyed by a bunch of conditional constructions dedicated to this logic. Maybe just put it all in one? E.g.:
c++
#ifdef QBT_USE_QCOLLATOR
// entire NaturalCompare declaration/definition
#else
// naturalCompare() function declaration
// entire NaturalCompare declaration/definition
#endif
```

Reply:
>Maybe just put it all in one? E

I pushed your first suggestion, it is slightly cleaner now. Let me know if you still think it is too messy.

Reply:
It looks a little better. But I would still prefer the second suggestion (if you don't mind). The motives are as follows:
1. The part of the code that you are trying to reuse, having a lot of conditional code, is very trivial for it to make enough sense.
2. The implementation of template methods is also very trivial and small for making sense to take it outside of the template declaration.

Reply:
Changed.

Reply:
>Personally I find it have some merits to separate the function/class declarations and definitions in a header file, i.e. it is easier to find what things are in it.

Generally, agree. But in case of such trivial function it only makes additional "noise", IMO.

Reply:
Last question.
Shouldn't these classes live in separate file? Header that contains only utility functions is included to .`cpp` files only. Now it is included in several other headers that is unlikely what we want, isn't it?

Reply:
I'm mostly fine with it and I suppose I can move it to "base/utils/compare.{cpp,h}" if you think it is better.

Reply:
>I can move it to "base/utils/compare.{cpp,h}" if you think it is better.

I think it's better.

Reply:
Done.